src/traces/choropleth/hover.js

```diff
@@ -80,12 +81,8 @@ function makeHoverInfo(pointData, trace, pt, axis) {
 
     if(hasZ) text.push(formatter(pt.z));
     if(hasText) {
-        var tx;
-
-        if(pt.tx) tx = pt.tx;
-        else if(trace.text) tx = trace.text;
-
-        if(!Array.isArray(tx)) text.push(tx);
+        var tx = Lib.extractOption(pt, trace, 'tx', 'text');
+        if(tx && tx !== '') text.push(tx);
```

Comment:
`''` is already falsey, right?

Reply:
Good catch. Strictly speaking, I think this should be:

```js
if(tx !== undefined && tx !== '') {
  text.push(tx);
}
```

as `Lib.extractOption` returns `undefined` for all falsy values except `0` in calcdata. But, `trace.text` is allowed to be `''` (which is the default by the way).

---

To make this easier to maintain (and remember :slightly_smiling_face: ), I'm thinking either 

- (1) we could remove the `dflt: ''` line in the `text` (and `hovertext`) attribute declarations. This leads to a slight change in behavior (but might more consistent) as now:

```js
{
   text: 'on-chart text',
   hovertext: ''
}
```

would use `''` for the hover label text field (i.e. not show anything). On `master`, `hovertext: ''` is ignored and `'on-chart text'` shows in the hover labels.

- (2) we could make `Lib.extractOption` return `undefined` when `trace.text: ''`, making it a little more opinionated (I think this only makes sense for string attributes).

Reply:
Here's what I came up with https://github.com/plotly/plotly.js/pull/2055/commits/3931273b331971085d39ce5bbb1fa7e9d4720c38

I made `Lib.extractOption` less opinionated, and made `scatter/fill_hover_text` handle falsy but valid text values.

Reply:
Yep, I like it. There are some edge cases this makes a little weird - eg trying to make one point have text but no hovertext, while all the others have both - but there are ways to get what you want (in that example using `hoverinfo` as an array) and I don't see any way to make *everything* more intuitive. So lets keep this, it's nice clean logic, will make sense to people, and will do what people expect 99.9% of the time.